python/src/opendp/extrinsics/_make_stateful_sequential_composition/__init__.py

```diff
@@ -0,0 +1,62 @@
+from opendp.measures import max_divergence
+from opendp.mod import Measurement, Transformation
+from opendp.core import new_user_queryable
+from opendp.measurements import make_user_measurement
+from opendp.extrinsics._utilities import to_then
+
+
+def make_stateful_sequential_composition(
+    input_domain, input_metric, privacy_measure, d_in, d_mids
+):
+    if privacy_measure != max_divergence(T=privacy_measure.distance_type):
+        raise ValueError("make_stateful_sequential_composition only supports pure-DP")
+
+    if d_in < 0:
+        raise ValueError("d_in must not be negative")
+    if any(d_i < 0 for d_i in d_mids):
+        raise ValueError("d_mids must not be negative")
+    
+    def function(arg):
+        def transition(query, _is_internal):
+            nonlocal input_domain, input_metric, arg, d_in, d_mids
+            d_mids = list(reversed(d_mids))
+
+            if query.input_space != (input_domain, input_metric):
+                raise ValueError(f"expected an input space of {(input_domain, input_metric)}, got {query.input_space}")
+
+            if isinstance(query, Measurement):
+                if query.output_measure != privacy_measure:
+                    raise ValueError(f"expected a privacy measure of {privacy_measure}, got {query.output_measure}")
+                if not query.check(d_in, d_mids[-1]):
+                    raise ValueError(f"query consumes {query.map(d_in)}, but is only allowed {d_mids[-1]}")
+
+                answer = query(arg)
+                d_mids.pop()
```

Comment:
I refactored the implementation to avoid the statefulness. Might be a cool feature to add to `make_sequential_composition`, but I've found I don't need it for either of what I thought were the two use-cases: this algorithm, and DPSGD.